### ISSUE TITLE
Use full URL (with cluster domain name) in admission webhook

### ISSUE THREAD

**Issue by antoineozenne** [kubernetes/kubernetes#107446]
### What happened?

My cluster is behind a corporate proxy. Using [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack), the apiserver raises an error when calling the mutating/validating admission webhook with a POST on `https://kube-prometheus-stack-operator.kube-prometheus-stack.svc:443/admission-prometheusrules/validate?timeout=10s`. This is because the wildcard `.svc` is not in the `NO_PROXY` environment variable in the apiserver. Instead, all my `NO_PROXY` environment variable contain the full cluster domain `.svc.cluster.local`, because this is used in the large majority of cases. 

### What did you expect to happen?

Can the apiserver use the full cluster domain name when calling webhooks? In my case, the URL would then be: `https://kube-prometheus-stack-operator.kube-prometheus-stack.svc.cluster.local:443/admission-prometheusrules/validate?timeout=10s`.

### How can we reproduce it (as minimally and precisely as possible)?

Just use an admission webhook in a cluster behind a corporate proxy.

### Anything else we need to know?

I know as a workaround, we can use `url` instead of `service` according to `https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#webhookclientconfig-v1-admissionregistration-k8s-io`, but it's less user-friendly and less flexible.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.11", GitCommit:"27522a29febbcc4badac257763044d0d90c11abd", GitTreeState:"clean", BuildDate:"2021-09-15T19:21:44Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.11", GitCommit:"27522a29febbcc4badac257763044d0d90c11abd", GitTreeState:"clean", BuildDate:"2021-09-15T19:16:25Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
hmm, I think that the problem is that the apiserver doesn't have access to the cluster.domain value, since it is configured on the kubelets only, but it will be good if somebody else can confirm this
```
--cluster-domain string
```

**Comment by aojea** [comment]
can you check your apiserver configuration for `enable-aggregator-routing` ? 
I think in your case it should be `false`, because if is `true`  I think the webhook will try to connect directly to the IP and don't do any dns resolution.
Can you confirm this?

**Comment by antoineozenne** [comment]
Yes, I confirm `enable-aggregator-routing` is not set, so `false`. Do I need to set this parameter? What are the other impacts/consequences?

**Comment by aojea** [comment]
@antoineozenne actually I can reproduce it.
A workaround is adding "svc" to NO_PROXY
I need to investigate more
/assign

**Comment by aojea** [comment]
This comment is just for reference:

1. apiserver initializes a resolver for use on the aggregated-apiserver and webhook

https://github.com/kubernetes/kubernetes/blob/8cc7d14588f688fec1d7b7da953e151158440b3a/cmd/kube-apiserver/app/server.go#L455

2. the resolver, depending on the `enable-aggregator-routing` variable, can resolve a service to an endpoint or to the clusterIP. The `kubernetes.default` service is always resolved to the current host

https://github.com/kubernetes/kubernetes/blob/8cc7d14588f688fec1d7b7da953e151158440b3a/cmd/kube-apiserver/app/server.go#L627-L644

3. webhooks initialize a client based on the configuration, and overrides the dialer with the custom resolver

https://github.com/kubernetes/kubernetes/blob/9c4905e7af0d070d139c06638ab9868fd3521883/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L161-L200

It seems (I have to confirm) golang uses the proxy on Transport, so the custom resolvers are never used because the Dialer receives the proxy URL

**Comment by fedebongio** [comment]
thanks @aojea for taking a look!
/triage accepted

**Comment by aojea** [comment]
I have to narrow down this function https://github.com/kubernetes/kubernetes/blob/45f77ca4bab08da88bdda267be2157dc10bc64ac/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L112-L114

**Comment by aojea** [comment]
> A workaround is adding "svc" to NO_PROXY

@antoineozenne there is a probability that removing the proxy from the webhooks no matter what can break some clusters
https://github.com/kubernetes/kubernetes/pull/107501#discussion_r783340420

is not possible to add `svc` to the NO_PROXY variable, `NO_PROXY=cluster.local,svc`

**Comment by antoineozenne** [comment]
@aojea thank you for your work! Yes, I have the possibility to add `.svc` to the `NO_PROXY` variable, but I would avoid that because in some environments (private providers), `.svc` is used in the infra...

If I understand, you proposed to bypass the proxy when using Services. It seems a bit much to me. Can we naively and simply use `.svc.<cluster domain>` (by providing it in the apiserver conf) to call webhooks?

**Comment by sftim** [comment]
> the apiserver doesn't have access to the cluster.domain value

This is what I believe too. Separately, it's a frustration because it also means that we can't provide the cluster domain name to a client via the API (because the API server doesn't have that information).

**Comment by chriswiggins** [comment]
Following up on this as we've just run into this. We want our webhook endpoints to have FQDN TLS certificates, but this restriction means we can only use URLs which is less than ideal.

Perhaps definitely something to plan in a future release as an option?

**Comment by aojea** [comment]
/unassign @aojea 

I do not have time to dedicate to this in the short term
